### PR TITLE
klog: Add WarningS

### PR DESCRIPTION
**Pull request by endocrimes**
This commit adds the missing `WarningS` function as an equivalent to
`InfoS` and `ErrorS`.

It currently cannot delegate to a logr logger as logr does not support
warning logs.

If we wanted to still delegate to an alternative logr implementation from `WarningS` like `InfoS` or `ErrorS` then we may want to consider pushing them to `logr.Info` as it's the closest approximation.

related to https://github.com/kubernetes/enhancements/issues/1602#issuecomment-715643825

**Release note**:

```release-note
Add support for structured warning logs via WarningS.

WarningS is similar to `InfoS` and `ErrorS`, but does not delegate to a custom `logr`.
```

### CODE REVIEW

**Comment by neolit123** [comment]
SGTM, i don't know why warningS was not part of the original PRs that added these - e.g.:
https://github.com/kubernetes/klog/pull/126

/assign yuzhiquan serathius

**Comment by dims** [comment]
Let's run this by @serathius :)

**Comment by torresdal** [comment]
Seams like `InfoS` replaces `Warning` as per migration instructions for structured logging, :

`klog.Warning, klog.Warningf, klog.Warningln, klog.WarningDepth -> klog.InfoS`

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#change-log-functions-to-structured-equivalent

**Comment by ehashman** [comment]
The migration instructions above contradicts the KEP, which explicitly states each `*f` function should be replaced with a `*S` equivalent:

https://github.com/kubernetes/enhancements/blame/69a0f01f66e764552691c1b3afe21b4718991587/keps/sig-instrumentation/1602-structured-logging/README.md#L103-L108

 I don't have a strong opinion in either direction, but IMO the KEP and migration plan should not say different things :)

File: klog.go [hunk 1]
```diff
@@ -1416,6 +1416,18 @@ func Warningf(format string, args ...interface{}) {
...
+// WarningS structured logs to the INFO and WARNING logs.
+// The msg argument used to add constant description to the log line.
+// The key/value pairs would be join by "=" ; a newline is always appended.
+//
+// Basic examples:
+// >> klog.WarningS("Page size is invalid", "page_size", "something", "error", err)
+// output:
+// >> W1025 00:15:15.525108       1 controller_utils.go:351] "Page size is invalid" page_size="something" error="test error"
+func WarningS(msg string, keysAndValues ...interface{}) {
+	logging.warningS(msg, keysAndValues...)
```
Comment (serathius) [inline klog.go:1428]: This change is incompatible with https://github.com/go-logr/logr interface and would not work with `SetLogger` interface.

**Comment by serathius** [comment]
I also don't have strong optinions, structured logging interface was based on https://github.com/go-logr/logr and  into Klog started before Structured Logging effort. I think the best person to answer if we need WarningS is Tim or Solly 
/assign @thockin @DirectXMan12

**Comment by endocrimes** [comment]
I don't have strong feelings here - (I spotted it while going to clean up some random logs) - my worry however is that not having like-for-like functions can potentially hinder the first step of cleaning up our existing logging. (Or at the very least spawn more people being confused like I was and going down the same rabbit hole, or assuming they should stay as Logf logs)

A review of existing format log -> structured log is fairly simple and unlikely to be controversial (and likely to include more useful context)

A review of someone changing a (or realistically many) log levels is less easy, because there's a decision to make at the same time as to if Info, V(n), or Error is most appropriate now.

That said, I missed the migration advice, so this might just be me being a little bit worried for nothing when linking that as context for initial didcussion.

**Comment by serathius** [comment]
If there are any problems with migrating/reviewing logs line changes then we should address them by updating Migration Guide https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md

**Comment by thockin** [comment]
So this gets pretty philosophical, borderline dogmatic.  Sorry to go here...

When we were putting together logr as an API, I was moved by http://dave.cheney.net/2015/11/05/lets-talk-about-logging.  Specifically the bit on warnings.  It resonated for me and aligned with my own experiences.  Every log-warning probably should have been an error or an info.  So I left warning out of logr, very intentionally, as a way to force the point.  It also (IMO) simplifies writing code - I don't have 3 choices, where 1 is really fuzzy.  I have 2 pretty stark choices.  For the code I have written against logr,  it has been successful.  One less thing to think about.  And not many people have complained.

As to this change - I do NOT think we should add klog APIs that don't map back to logr.  That seems like a bomb waiting to go off.  We *could* add WarningS that maps to logr.Info, but it seems like a guess - at that point, why bother?  Force the call-site to decide Info vs Error.  It's not a binding decision, it can always be changed later...

So for my take on this - I would say NAK on adding WarningS, even with the acknowledgement that a human needs to make a decision for those callsites.  I count about 900 of them (vs 8000+ Info).  I think it's probably safe to just convert them to V(0) or plain Info (same thing) and let special cases get caught in reviews.

Or someone could go thru all of them and convert to Info/Error in bulk... fun...

**Comment by dims** [comment]
/hold

thanks for the feedback @thockin. will wait for a bit before we close this out.

**Comment by endocrimes** [comment]
@thockin Sounds good to me (fwiw I mostly agree on the philosophical front in most cases). Given that, and where we want to go, do we want to consider officially deprecating the warnings API in klog entirely?

**Comment by dims** [comment]
@endocrimes +1 to deprecate Warnings API

**Comment by DirectXMan12** [comment]
> Given that, and where we want to go, do we want to consider officially deprecating the warnings API in klog entirely

This seems pretty reasonable to me, specially since a `Deprecated: XYZ` notice can be checked by tooling so folks can get warnings (heh) from linters and such.

**Pull request closed without merging**